crates/turborepo-lockfiles/src/npm.rs

```diff
@@ -0,0 +1,365 @@
+use std::collections::HashMap;
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+use super::{Error, Lockfile, Package};
+
+// we change graph traversal now
+// resolve_package should only be used now for converting initial contents
+// of workspace package.json into a set of node ids
+#[derive(Debug, Serialize, Deserialize)]
+pub struct NpmLockfile {
+    #[serde(rename = "lockfileVersion")]
+    lockfile_version: i32,
+    packages: HashMap<String, NpmPackage>,
+    // We parse this so it doesn't end up in 'other' and we don't need to worry
+    // about accidentally serializing it.
+    #[serde(skip_serializing, default)]
+    dependencies: HashMap<String, Value>,
+    // We want to reserialize any additional fields, but we don't use them
+    // we keep them as raw values to avoid describing the correct schema.
+    #[serde(flatten)]
+    other: HashMap<String, Value>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct NpmPackage {
+    version: Option<String>,
+    resolved: Option<String>,
+    #[serde(default)]
+    dependencies: HashMap<String, String>,
+    #[serde(default)]
+    dev_dependencies: HashMap<String, String>,
+    #[serde(default)]
+    peer_dependencies: HashMap<String, String>,
+    #[serde(default)]
+    optional_dependencies: HashMap<String, String>,
+    // We want to reserialize any additional fields, but we don't use them
+    // we keep them as raw values to avoid describing the correct schema.
+    #[serde(flatten)]
+    other: HashMap<String, Value>,
+}
+
+impl Lockfile for NpmLockfile {
+    fn resolve_package(
+        &self,
+        workspace_path: &str,
+        name: &str,
+        _version: &str,
+    ) -> Result<Option<Package>, Error> {
+        if !self.packages.contains_key(workspace_path) {
+            return Err(Error::MissingWorkspace(workspace_path.to_string()));
+        }
+
+        let possible_keys = [
+            // AllDependencies will return a key to avoid choosing the incorrect transitive dep
+            name.to_string(),
+            // If we didn't find the entry just using name, then this is an initial call to
+            // ResolvePackage based on information coming from internal packages'
+            // package.json First we check if the workspace uses a nested version of
+            // the package
+            format!("{}/node_modules/{}", workspace_path, name),
+            // Next we check for a top level version of the package
+            format!("node_modules/{}", name),
+        ];
+        possible_keys
+            .into_iter()
+            .filter_map(|key| {
+                self.packages.get(&key).map(|pkg| {
+                    let version = pkg.version.clone().unwrap_or_default();
+                    Ok(Package { key, version })
+                })
+            })
+            .next()
+            .transpose()
+    }
+
+    fn all_dependencies(&self, key: &str) -> Result<Option<HashMap<String, &str>>, Error> {
```

Comment:
Does it make sense to not transpose here? That way the top level option defines the existence of the key, rather than the alternative which is 'either an optional key or one of the deps is missing a version'. Something like this:

```suggestion

    /// Gets all the dependencies of a package, if it exists in the graph.
    /// If None, the package key does not exist.
    /// If Some(Error), the package exists, but its dependencies are invalid.
    fn all_dependencies(&self, key: &str) -> Option<Result<Hashmap<String, &str>, Error>> {
```

Reply:
This is a slight oddity coming from trying to keep this comparable to Go so we can ship changes to both implementations while we're supporting both. In reality, this should be a panic as a version will only be missing if an external package depends on a workspace. I don't want to panic as this is getting called via FFI and would rather let Go handle the error.

Once we stop maintaining/shipping the Go version of the lockfiles there's a layer of Rust we should apply to these implementations as we can rely on types more, but I think that should wait until we're down to one implementation.